File: app/src/main/java/hu/kts/cmetronome/ui/workout/WorkoutController.java
package hu.kts.cmetronome.ui.workout;

import android.app.Activity;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import hu.kts.cmetronome.R;
import hu.kts.cmetronome.Settings;
import hu.kts.cmetronome.Sounds;
import hu.kts.cmetronome.WorkoutStatus;
import hu.kts.cmetronome.repository.WorkoutRepository;

public class WorkoutController implements LifecycleObserver {

    @BindView(R.id.rep_counter)
    TextView repCounterTextView;
    @BindView(R.id.set_counter)
    TextView setCounterTextView;

    private Activity activity;
    private final WorkoutRepository repository;

    private Stopwatch stopWatch;
    private Sounds sounds;
    private IndicatorAnimation indicatorAnimation;
    private Settings settings;
    private Help help;
    private Countdowner countdowner;

    public WorkoutController(Activity activity, WorkoutRepository workoutRepository, Settings settings) {
        this.activity = activity;
        this.repository = workoutRepository;
        this.settings = settings;
        ButterKnife.bind(this, activity);
        help = new Help(activity);
        sounds = new Sounds(activity);
        countdowner = new Countdowner(activity, this::startWorkout, this::onCountdownCancelled, settings);
        stopWatch = new Stopwatch(activity);
        initWorkoutData();
        initSettingsRelatedParts();

    }

    private void initWorkoutData() {
        if (repository.getWorkoutStatus() != null) {
            WorkoutStatus savedWorkoutStatus = repository.getWorkoutStatus();
            setWorkoutStatusAndHelpText(savedWorkoutStatus == WorkoutStatus.IN_PROGRESS || savedWorkoutStatus == WorkoutStatus.COUNTDOWN_IN_PROGRESS? WorkoutStatus.PAUSED : savedWorkoutStatus);
            if (repository.getWorkoutStatus() == WorkoutStatus.BETWEEN_SETS) {
                stopWatch.start(repository.getStopwatchStartTime());
            }
        } else {
            setWorkoutStatusAndHelpText(WorkoutStatus.BEFORE_START);
        }
        fillRepCounterTextViewWithTruncatedData();
        fillSetCounterTextViewWithTruncatedData();
    }

    public void onRepCounterClick() {
        switch (repository.getWorkoutStatus()) {
            case BEFORE_START:
            case PAUSED:
            case BETWEEN_SETS: countDownAndStart(); break;
            case COUNTDOWN_IN_PROGRESS:
            case IN_PROGRESS: pauseWorkout(); break;
        }
    }

    public boolean onRepCounterLongClick() {
        if (repository.getWorkoutStatus() == WorkoutStatus.IN_PROGRESS || repository.getWorkoutStatus() == WorkoutStatus.PAUSED) {
            stopSet();
            return true;
        } if (repository.getWorkoutStatus() == WorkoutStatus.BETWEEN_SETS) {
            resetWorkout();
            return true;
        }
        return false;
    }

    public void pauseWorkout() {
        setWorkoutStatusAndHelpText(WorkoutStatus.PAUSED);
        countdowner.cancel();
        sounds.stop();
        resetIndicator();
    }

    private void onCountdownCancelled() {
        fillRepCounterTextViewWithTruncatedData();
    }

    private void resetIndicator() {
        getIndicatorAnimation().stop();
    }


    private void countDownAndStart() {
        if (repository.getWorkoutStatus() == WorkoutStatus.BETWEEN_SETS) {
            repository.resetRepCounter();
        }
        stopWatch.stop();
        setWorkoutStatusAndHelpText(WorkoutStatus.COUNTDOWN_IN_PROGRESS);
        countdowner.start();
    }

    private void startWorkout() {
        setWorkoutStatusAndHelpText(WorkoutStatus.IN_PROGRESS);
        fillRepCounterTextViewWithTruncatedData();
        getIndicatorAnimation().start();
    }

    /**
     * Initialization can't be called from onCreate because animation needs the actual size of the elements
     */
    private IndicatorAnimation getIndicatorAnimation() {
        if (indicatorAnimation == null) {
            indicatorAnimation = new IndicatorAnimation(activity, this::onIndicatorAnimationEvent);
        }
        return indicatorAnimation;
    }

    private void onIndicatorAnimationEvent(IndicatorAnimation.Event event) {
        switch (event) {
            case DOWN:
                sounds.makeUpSound();
                break;
            case UP:
                sounds.makeDownSound();
                break;
            case LEFT:
                repository.increaseRepCounter();
                fillRepCounterTextViewWithTruncatedData();
                break;
        }
    }

    private void stopSet() {
        setWorkoutStatusAndHelpText(WorkoutStatus.BETWEEN_SETS);
        resetIndicator();
        sounds.stop();
        countdowner.cancel();
        stopWatch.start();
        repository.setStopwatchStartTime(stopWatch.getStartTime());
        increaseSetCounter();
    }

    private void setWorkoutStatusAndHelpText(WorkoutStatus status) {
        repository.setWorkoutStatus(status);
        help.setHelpTextByWorkoutStatus(status);
    }

    private void increaseSetCounter() {
        repository.increaseSetCounter();
        fillSetCounterTextViewWithTruncatedData();
    }


    private void resetWorkout() {
        setWorkoutStatusAndHelpText(WorkoutStatus.BEFORE_START);
        stopWatch.stop();
        repository.resetCounters();
        fillSetCounterTextViewWithTruncatedData();
        fillRepCounterTextViewWithTruncatedData();
    }

    public void initSettingsRelatedParts() {
        help.setEnabled(settings.isShowHelp());
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void onDestroy() {
        sounds.release();
        stopWatch.stop();
        countdowner.cancel();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    public void onPause() {
        if (repository.getWorkoutStatus() == WorkoutStatus.IN_PROGRESS || repository.getWorkoutStatus() == WorkoutStatus.COUNTDOWN_IN_PROGRESS) {
            pauseWorkout();
        }
    }

    /**
     * counter should contain maximum 2 digits
     */
    private void fillRepCounterTextViewWithTruncatedData() {
        repCounterTextView.setText(String.valueOf(repository.getRepCount() % 100));
    }

    /**
     * counter should contain maximum 2 digits
     */
    private void fillSetCounterTextViewWithTruncatedData() {
        setCounterTextView.setText(String.valueOf(repository.getSetCount() % 100));
    }
}

File: app/src/main/java/hu/kts/cmetronome/ui/workout/WorkoutController.kt
package hu.kts.cmetronome.ui.workout

import android.content.SharedPreferences
import android.util.Log
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import hu.kts.cmetronome.R
import hu.kts.cmetronome.TimeProvider
import hu.kts.cmetronome.WorkoutStatus
import hu.kts.cmetronome.di.TimeProviderRep
import hu.kts.cmetronome.logic.WorkoutCalculations
import hu.kts.cmetronome.repository.WorkoutRepository
import hu.kts.cmetronome.repository.WorkoutSettings
import hu.kts.cmetronome.sounds.Sounds
import hu.kts.cmetronome.ui.Toaster
import javax.inject.Inject

class WorkoutController @Inject constructor(private val repository: WorkoutRepository,
                                            private val settings: WorkoutSettings,
                                            private val sounds: Sounds,
                                            @TimeProviderRep private val timeProviderRep: TimeProvider,
                                            private val calculations: WorkoutCalculations,
                                            private val indicatorAnimation: IndicatorAnimation,
                                            private val stopWatch: Stopwatch,
                                            private val help: Help,
                                            private val countdowner: Countdowner,
                                            private val counters: Counters,
                                            private val toaster: Toaster
) : DefaultLifecycleObserver {

    //we have to hold a reference to this or else it'd be gc-d
    private val listener: SharedPreferences.OnSharedPreferenceChangeListener = SharedPreferences.OnSharedPreferenceChangeListener { _, key -> onSettingsChanged(key) }

    init {
        initWorkoutData()
        timeProviderRep.observeForever { count -> onRepTimeProviderTick(count) }
        settings.addListener(listener)
        help.setEnabled(settings.isShowHelp)
        countdowner.onFinish = this::startWorkout
        countdowner.onCancel = this::onCountdownCancelled
    }

    private fun initWorkoutData() {
        if (repository.workoutStatus != null) {
            val savedWorkoutStatus = repository.workoutStatus
            setWorkoutStatusAndHelpText(if (savedWorkoutStatus == WorkoutStatus.IN_PROGRESS || savedWorkoutStatus == WorkoutStatus.COUNTDOWN_IN_PROGRESS) WorkoutStatus.PAUSED else savedWorkoutStatus!!)
            if (repository.workoutStatus == WorkoutStatus.BETWEEN_SETS) {
                stopWatch.start(repository.stopwatchStartTime)
            }
        } else {
            setWorkoutStatusAndHelpText(WorkoutStatus.BEFORE_START)
        }
        counters.setRepCount(repository.repCount)
        counters.setSetCount(repository.setCount)
    }

    private fun onSettingsChanged(key: String?) {
        when (key) {
            WorkoutSettings.KEY_SHOW_HELP -> help.setEnabled(settings.isShowHelp)
            WorkoutSettings.KEY_REP_STARTS_WITH_UP -> resetIndicator()
        }
    }

    fun onRepCounterClick() {
        when (repository.workoutStatus) {
            WorkoutStatus.BEFORE_START, WorkoutStatus.PAUSED, WorkoutStatus.BETWEEN_SETS -> countDownAndStart()
            WorkoutStatus.COUNTDOWN_IN_PROGRESS, WorkoutStatus.IN_PROGRESS -> pauseWorkout()
        }
    }

    fun onRepCounterLongClick(): Boolean {
        if (repository.workoutStatus == WorkoutStatus.IN_PROGRESS || repository.workoutStatus == WorkoutStatus.PAUSED) {
            stopSet()
            return true
        }
        if (repository.workoutStatus == WorkoutStatus.BETWEEN_SETS) {
            resetWorkout()
            return true
        }
        return false
    }

    private fun pauseWorkout() {
        setWorkoutStatusAndHelpText(WorkoutStatus.PAUSED)
        countdowner.cancel()
        sounds.stop()
        resetIndicator()
    }

    private fun onCountdownCancelled() {
        counters.setRepCount(repository.repCount)
    }

    private fun resetIndicator() {
        indicatorAnimation.stop()
        timeProviderRep.stop()
    }


    private fun countDownAndStart() {
        if (settings.repDownTime + settings.repUpTime == 0L) {
            toaster.showShort(R.string.rep_is_empty_message)
            return
        }
        if (repository.workoutStatus == WorkoutStatus.BETWEEN_SETS) {
            repository.resetRepCounter()
        }
        stopWatch.stop()
        setWorkoutStatusAndHelpText(WorkoutStatus.COUNTDOWN_IN_PROGRESS)
        countdowner.start()
    }

    private fun startWorkout() {
        setWorkoutStatusAndHelpText(WorkoutStatus.IN_PROGRESS)
        counters.setRepCount(repository.repCount)
        timeProviderRep.startUp()
    }

    private fun onRepTimeProviderTick(count: Long) {
        calculations.getNextDirection(count)?.let {
            Log.d(TAG, "${count / 2} -> $it")
            when (it) {
                IndicatorAnimation.Direction.DOWN -> {
                    sounds.makeDownSound()
                    indicatorAnimation.start(IndicatorAnimation.Direction.DOWN)
                }
                IndicatorAnimation.Direction.RIGHT ->
                    indicatorAnimation.start(IndicatorAnimation.Direction.RIGHT)
                IndicatorAnimation.Direction.UP -> {
                    sounds.makeUpSound()
                    indicatorAnimation.start(IndicatorAnimation.Direction.UP)
                }
                IndicatorAnimation.Direction.LEFT -> {
                    indicatorAnimation.start(IndicatorAnimation.Direction.LEFT)
                }
            }

        }
        if (calculations.shouldIncreaseRepCounter(count)) {
            repository.increaseRepCounter()
            counters.setRepCount(repository.repCount)
        }
    }

    private fun stopSet() {
        setWorkoutStatusAndHelpText(WorkoutStatus.BETWEEN_SETS)
        resetIndicator()
        sounds.stop()
        countdowner.cancel()
        stopWatch.start()
        repository.stopwatchStartTime = stopWatch.startTime
        increaseSetCounter()
    }

    private fun setWorkoutStatusAndHelpText(status: WorkoutStatus) {
        repository.workoutStatus = status
        help.setHelpTextByWorkoutStatus(status)
    }

    private fun increaseSetCounter() {
        repository.increaseSetCounter()
        counters.setSetCount(repository.setCount)
    }


    private fun resetWorkout() {
        setWorkoutStatusAndHelpText(WorkoutStatus.BEFORE_START)
        stopWatch.stop()
        repository.resetCounters()
        counters.setSetCount(repository.setCount)
        counters.setRepCount(repository.repCount)
    }

    override fun onDestroy(owner: LifecycleOwner) {
        stopWatch.stop()
        countdowner.cancel()
    }

    override fun onPause(owner: LifecycleOwner) {
        if (repository.workoutStatus == WorkoutStatus.IN_PROGRESS || repository.workoutStatus == WorkoutStatus.COUNTDOWN_IN_PROGRESS) {
            pauseWorkout()
        }
    }

    companion object {
        const val TAG = "WorkoutController"
    }

}
